lib/stats_by_agency_route_stop_hour.js

```diff
@@ -0,0 +1,41 @@
+'use strict'
+
+const afterAll = (opt) => {
+	let materialized = false
+	if (opt.statsByAgencyIdAndRouteIdAndStopAndHour === 'materialized-view') {
+		materialized = true
+	} else if (opt.statsByAgencyIdAndRouteIdAndStopAndHour !== 'view') {
+		throw new Error('invalid opt.statsByAgencyIdAndRouteIdAndStopAndHour, must be one of these: none, view, materialized-view.')
+	}
+	const createViewCmd = materialized
+		? `CREATE MATERIALIZED VIEW`
+		: `CREATE OR REPLACE VIEW`
+
+	return `\
+${createViewCmd} "${opt.schema}".stats_by_agency_route_stop_hour AS
+SELECT DISTINCT ON (agency_id, route_id, stop_id, effective_hour)
+	agency_id, route_id, stop_id, station_id,
+	"date" as service_date,
+	date_trunc('hour', t_arrival) AS effective_hour,
+	count(*) OVER (PARTITION BY route_id, stop_id, date_trunc('hour', t_arrival)) AS nr_of_arrs
+FROM "${opt.schema}".arrivals_departures;
+
+${materialized ? `\
+CREATE INDEX ON "${opt.schema}".stats_by_agency_route_stop_hour (route_id);
+CREATE INDEX ON "${opt.schema}".stats_by_agency_route_stop_hour (stop_id);
+CREATE INDEX ON "${opt.schema}".stats_by_agency_route_stop_hour (effective_hour);
+CREATE INDEX ON "${opt.schema}".stats_by_agency_route_stop_hour (route_id, stop_id);
+CREATE INDEX ON "${opt.schema}".stats_by_agency_route_stop_hour (route_id, effective_hour);
+CREATE INDEX ON "${opt.schema}".stats_by_agency_route_stop_hour (stop_id, effective_hour);
+CREATE INDEX ON "${opt.schema}".stats_by_agency_route_stop_hour (route_id, stop_id, effective_hour);
```

Comment:
I'd assume stop_id to be more specific than route_id, and give reordering a try (using stop_id, route_id, effective_hour)

Reply:
With the 2023-05-02 VBB GTFS, the indexes defined above take up ~7GB, while the materialised view itself takes up ~10GB.

The [PostgreSQL docs say this](https://www.postgresql.org/docs/15/indexes-multicolumn.html):

> A multicolumn B-tree index can be used with query conditions that involve any subset of the index's columns, but the index is most efficient when there are constraints on the leading (leftmost) columns.

However, it seems like having (a) multi-column index(es) is not necessary, as filtering by one of the columns (`route_id` or `stop_id` or `station_id` or `effective_hour`) reduces the number of rows enough (down to 10^4 - 10^5) for queries to be fast enough (<= 10ms). Defining one single-column index for each column yields ~3GB of indexes.

What do you think?